File: gaia-sdk-javascript/src/api/SkillRef.ts
import {defer, Observable} from 'rxjs';
import {GaiaStreamClient} from '../graphql/GaiaStreamClient';
import {flatMap, map} from 'rxjs/operators';

export class SkillRef {
    private readonly client: GaiaStreamClient;
    private readonly uri: string;

    constructor(uri: string, client: GaiaStreamClient) {
        this.uri = uri;
        this.client = client;
    }

    public start(): Observable<{}> {
        return defer(() => this.client.post({uri: this.uri}, '/skill/start'));
    }

    public stop(): Observable<{}> {
        return defer(() => this.client.post({uri: this.uri}, '/skill/stop'));
    }

    public logs(numberOfLines?: number): Observable<string> {
        return this.logsInternal(numberOfLines)
            .pipe(flatMap(response => defer(() => response.logLines)));
    }

    private logsInternal(numberOfLines?: number): Observable<SkillProvisionLogs> {
        return defer(() => this.client.post({uri: this.uri, numberOfLines: numberOfLines}, '/skill/logs'));
    }

    public build(): Observable<SkillProvisionBuildResponse> {
        return defer(() => this.client.post({uri: this.uri}, '/skill/build'));
    }

    public evaluate(contract: string, payload: any): Observable<SkillEvaluation>;
    public evaluate(payload: any): Observable<SkillEvaluation>;
    public evaluate(first: any, second?: any): Observable<SkillEvaluation> {
        if (typeof first === 'string' && second !== undefined) {
            //corresponds to evaluate(contract, payload)
            return this.doEvaluate(first, second);
        }
        if (typeof first !== 'string' && second === undefined) {
            // corresponds to evaluate(payload)
            return this.doEvaluate(undefined, first);
        }
        throw new Error('Either use evaluate(contract:string, payload:{}) or evaluate(payload: {})');
    }

    private doEvaluate(contract: string | undefined, p: any):  Observable<SkillEvaluation> {
        if (!p) throw new Error('You must provide a payload');
        const request = {uri: this.uri, payload: p};
        if (contract) {
            request['contract'] = contract;
        }
        return defer(() => this.client.post(request, '/skill/evaluate'))
            .pipe(map(r => new SkillEvaluation(r)));
    }
}

export class SkillProvisionBuildResponse {
    reference: string;

    constructor(reference: string) {
        this.reference = reference;
    }
}
export class SkillProvisionStatus {
    name: string;
    status: string;
    createdAt: string;

    constructor(name: string, status: string, createdAt: string) {
        this.name = name;
        this.status = status;
        this.createdAt = createdAt;
    }
}

export class SkillProvisionLogs {
    logLines: [string];

    constructor(logLines: [string]) {
        this.logLines = logLines;
    }
}

export class SkillEvaluation {

    private readonly _response: any;

    constructor(payload: any) {
        this.response = payload;
    }

    response(): any {
        return this._response;
    }
}

File: gaia-sdk-python/gaia_sdk/api/SkillRef.py
import json

import rx
import rx.operators as ops
from rx.core.typing import Observable, Scheduler

from gaia_sdk.http import GaiaStreamClient


class SkillProvisionCanceledResponse:
    def __init__(self, reference: str):
        self._reference = reference

    @property
    def reference(self) -> str:
        return self._reference


class SkillBuildResponse:
    def __init__(self, reference):
        self._reference = reference

    @property
    def reference(self) -> str:
        return self._reference


class SkillEvaluation:
    def __init__(self, response: dict):
        self._response = response

    @property
    def response(self):
        return self._response


class SkillRef:
    def __init__(self, uri: str, client: GaiaStreamClient, scheduler: Scheduler):
        self._uri = uri
        self._client = client
        self._scheduler = scheduler

    def start(self) -> Observable[any]:
        return rx.from_callable(lambda: self._client.post_json({'uri': self._uri}, url_postfix="/skill/start"),
                                self._scheduler) \
            .pipe(
            ops.map(lambda r: {})
        )

    def stop(self) -> Observable[any]:
        return rx.from_callable(lambda: self._client.post_json({'uri': self._uri}, url_postfix="/skill/stop"),
                                self._scheduler) \
            .pipe(
            ops.map(lambda r: {})
        )

    def logs(self, number_of_lines: int) -> Observable[str]:
        return rx.from_callable(lambda: self._client.post_json({'uri': self._uri, 'numberOfLines': number_of_lines},
                                                               url_postfix="/skill/logs"), self._scheduler) \
            .pipe(
            ops.map(lambda r: json.loads(r.content)),
            ops.flat_map(lambda r: rx.from_iterable(r['logLines']))
        )

    def cancel(self) -> Observable[SkillProvisionCanceledResponse]:
        return rx.from_callable(lambda: self._client.post_json({'uri': self._uri}, url_postfix="/skill/cancel"),
                                self._scheduler) \
            .pipe(
            ops.map(lambda r: json.loads(r.content)),
            ops.map(lambda r: SkillProvisionCanceledResponse(r["reference"]))
        )

    def build(self) -> Observable[SkillBuildResponse]:
        return rx.from_callable(lambda: self._client.post_json({'uri': self._uri}, url_postfix="/skill/build"),
                                self._scheduler) \
            .pipe(
            ops.map(lambda r: json.loads(r.content)),
            ops.map(lambda r: SkillBuildResponse(r["reference"]))
        )

    def evaluate(self, payload: dict, contract: str = None) -> Observable[SkillEvaluation]:
        request = {'uri': self._uri, 'payload': payload}
        if contract is not None:
            request['contract'] = contract
        return rx.from_callable(
            lambda: self._client.post_json(request, url_postfix="/skill/evaluate"),
            self._scheduler) \
            .pipe(
            ops.map(lambda response: json.loads(response.content)),
            ops.map(lambda d: SkillEvaluation(d)))
